[wasm] Fix fragmentation caused by large objects allocation

BrzVlad opened the pull request:
For large objects (bigger than 8K), we have two different allocators. We either allocate them in LOS sections (a section has 1MB), if the object is smaller than 1MB in size, or allocate it directly from the OS. A LOS section has 1MB alignment. On wasm this huge alignment is problematic leading to memory fragmentation. This PR disables allocation from LOS sections, allocating each object independently.

Testing allocation for 10k sized objects with `console-v8` this change showed the following results
- allocation speed is about 20% faster
- fragmentation becomes 0. Data obtained from `GC.GetGCMemoryInfo`:
   - BEFORE: Total Memory: 99.61 MB | HeapSizeBytes: 236.59 MB | FragmentedBytes: 141.00 MB
   - AFTER: Total Memory: 99.61 MB | HeapSizeBytes: 96.21 MB | FragmentedBytes: 634.73 KB

Fixes https://github.com/dotnet/runtime/issues/118044

BrzVlad committed:
[wasm] Disable allocation from los sections

Allocation in the segments is normally done for objects smaller than 1MB. The problem is that for los sections we require 1MB alignment and on wasm we end up allocating twice the memory in order to ensure the alignment. With this commit we reuse the same allocation code that was used for objects bigger than 1MB. This still uses the unfortunate alignment of page size.

BrzVlad committed:
[wasm] Avoid page size alignment for LOS objects smaller than LOS_SECTION_OBJECT_LIMIT

On mono, objects need to have the alignment of 8 bytes. We allocate using this alignemnt directly for objects smaller than 1MB. This further decreases fragmentation and increases allocation speed.

pavelsavara commented:
[Log](https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-118099-merge-0df032ac924d4c398b/WasmTestOnChrome-ST-System.Text.RegularExpressions.Tests.Attempt.2/1/console.e396bb11.log?helixlogtype=result&skoid=8eda00af-b5ec-4be9-b69b-0919a2338892&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2025-07-28T09%3A21%3A04Z&ske=2025-07-28T10%3A21%3A04Z&sks=b&skv=2024-11-04&sv=2024-11-04&se=2025-07-28T10%3A21%3A04Z&sr=b&sp=rl&sig=Kx8U5HR%2FZDLVqg5BzK8jDT%2BzBQsB3tZr%2FRZ%2Fj2CkjME%3D)
```
[21:06:27] fail: [MONO] no object of size 3879832992d
                 
                 Error
                     at Mc (http://127.0.0.1:36981/_framework/dotnet.runtime.js:3:171883)
                     at wasm_trace_logger (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[158]:0xb02e)
                     at eglib_log_adapter (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[656]:0x455b0)
                     at monoeg_g_logv_nofree (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[581]:0x436a5)
                     at monoeg_g_log (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[583]:0x43768)
                     at ms_find_block_obj_size_index (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[952]:0x52e56)
                     at alloc_obj (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[991]:0x59399)
                     at major_alloc_object (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[982]:0x54a25)
                     at alloc_for_promotion (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[1058]:0x5c4c7)
                     at copy_object_no_checks (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[998]:0x5a036)
                     at drain_gray_stack (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[987]:0x5609f)
                     at finish_gray_stack (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[903]:0x506ff)
                     at major_do_collection (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[879]:0x4e6cd)
                     at sgen_perform_collection (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[876]:0x4dbfa)
                     at sgen_ensure_free_space (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[875]:0x4daaa)
                     at sgen_los_alloc_large_inner (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[948]:0x52b7f)
                     at sgen_alloc_obj_nolock (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[861]:0x4ce21)
                     at mono_gc_alloc_vector (http://127.0.0.1:36981/_framework/dotnet.native.wasm:wasm-function[3060]:0xec9e4)
```

[Log](https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-118099-merge-0df032ac924d4c398b/WasmTestOnChrome-ST-System.Text.RegularExpressions.Unit.Tests/1/console.ff4cd2b3.log?helixlogtype=result&skoid=8eda00af-b5ec-4be9-b69b-0919a2338892&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2025-07-28T09%3A21%3A07Z&ske=2025-07-28T10%3A21%3A07Z&sks=b&skv=2024-11-04&sv=2024-11-04&se=2025-07-28T10%3A21%3A07Z&sr=b&sp=rl&sig=DbUg75CnhVdefYke5zaby05OgrnjNL%2FghQDV03WZeIM%3D)
```
[21:01:52] fail: [MONO] * Assertion: should not be reached at /__w/1/s/src/mono/mono/sgen/sgen-scan-object.h:93
                 
                 Error
                     at Mc (http://127.0.0.1:35905/_framework/dotnet.runtime.js:3:171883)
                     at wasm_trace_logger (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[158]:0xb02e)
                     at eglib_log_adapter (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[656]:0x455b0)
                     at monoeg_g_logv_nofree (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[581]:0x436a5)
                     at monoeg_assertion_message (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[584]:0x43796)
                     at mono_assertion_message_unreachable (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[586]:0x437f0)
                     at simple_nursery_serial_scan_object (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[1062]:0x5d557)
                     at simple_nursery_serial_drain_gray_stack (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[1059]:0x5c532)
                     at finish_gray_stack (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[903]:0x506ff)
                     at collect_nursery (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[878]:0x4e33e)
                     at sgen_perform_collection (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[876]:0x4db3a)
                     at sgen_ensure_free_space (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[875]:0x4daaa)
                     at sgen_alloc_obj_nolock (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[861]:0x4cf45)
                     at mono_gc_alloc_obj (http://127.0.0.1:35905/_framework/dotnet.native.wasm:wasm-function[3054]:0xec6ff)
```

[Log](https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-118099-merge-0df032ac924d4c398b/WasmTestOnChrome-ST-System.Data.Common.Tests/1/console.31ef2e27.log?helixlogtype=result&skoid=8eda00af-b5ec-4be9-b69b-0919a2338892&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2025-07-28T09%3A20%3A54Z&ske=2025-07-28T10%3A20%3A54Z&sks=b&skv=2024-11-04&sv=2024-11-04&se=2025-07-28T10%3A20%3A54Z&sr=b&sp=rl&sig=Pzoge11Gz9gfLFMF%2BjvUyN6kRSyV4SLki9otLg0AoJI%3D)
```
[21:00:15] info: Starting:    System.Data.Common.Tests.dll
[21:00:18] fail: MONO_WASM: memory access out of bounds
                 RuntimeError: memory access out of bounds
                     at mono_property_bag_get (http://127.0.0.1:45891/_framework/dotnet.native.wasm:wasm-function[2657]:0xd3187)
                     at mono_error_set_for_class_failure (http://127.0.0.1:45891/_framework/dotnet.native.wasm:wasm-function[1186]:0x67d03)
                     at mono_class_is_assignable_from_general (http://127.0.0.1:45891/_framework/dotnet.native.wasm:wasm-function[1240]:0x69f9a)
                     at mono_class_is_assignable_from_checked (http://127.0.0.1:45891/_framework/dotnet.native.wasm:wasm-function[1236]:0x69cf7)
                     at mono_interp_isinst (http://127.0.0.1:45891/_framework/dotnet.native.wasm:wasm-function[211]:0xdd1e)
```

BrzVlad committed:
Fix mapping of the same card to 2 different los objects

This will produce some waste, but 512 alignment is still quite small compared to the size of the objects here (> 8k)

pavelsavara commented:
/azp run runtime-wasm

pavelsavara reviewed (approved):
I guess that objects on LOS are never moved, so from this perspective it's no change and no perf impact. 

I wonder if there is a allocation pattern for which this will cause more fragmentation than the previous state.

cc @kg

pavelsavara commented:
/ba-g all the CI failures are unrelated
